cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/LeaderReplicaDistributionGoal.java

```diff
@@ -0,0 +1,354 @@
+                                           ActionType.LEADERSHIP_MOVEMENT,
+                                           optimizedGoals,
+                                           optimizationOptions);
+      // Only check if we successfully moved something.
+      if (b != null) {
+        if (broker.leaderReplicas().size() <= balanceUpperLimit) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  private boolean rebalanceByMovingLeadershipIn(Broker broker,
+                                                ClusterModel clusterModel,
+                                                Set<Goal> optimizedGoals,
+                                                OptimizationOptions optimizationOptions) {
+    if (optimizationOptions.excludedBrokersForLeadership().contains(broker.id())) {
+      return true;
+    }
+
+    Set<Broker> candidateBrokers =  Collections.singleton(broker);
+    for (Replica replica : broker.replicas()) {
+      if (replica.isLeader()) {
+        continue;
+      }
+      Broker b = maybeApplyBalancingAction(clusterModel,
+                                           clusterModel.partition(replica.topicPartition()).leader(),
+                                           candidateBrokers,
+                                           ActionType.LEADERSHIP_MOVEMENT,
+                                           optimizedGoals,
+                                           optimizationOptions);
+      // Only check if we successfully moved something.
+      if (b != null) {
+        if (broker.leaderReplicas().size() >= _balanceLowerLimit) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  private boolean rebalanceByMovingLeaderReplicasOut(Broker broker,
+                                                    ClusterModel clusterModel,
+                                                    Set<Goal> optimizedGoals,
+                                                    OptimizationOptions optimizationOptions) {
+    // Get the eligible brokers.
+    SortedSet<Broker> candidateBrokers = new TreeSet<>(Comparator.comparingInt((Broker b) -> b.leaderReplicas().size())
+                                                                 .thenComparingInt(Broker::id));
+    candidateBrokers.addAll(_selfHealingDeadBrokersOnly ? clusterModel.aliveBrokers()
+                                                        : clusterModel.aliveBrokers()
+                                                                      .stream()
+                                                                      .filter(b -> b.leaderReplicas().size() < _balanceUpperLimit)
+                                                                      .collect(Collectors.toSet()));
+
+    int balanceUpperLimit = broker.isAlive() ? _balanceUpperLimit : 0;
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+    for (Replica replica : new HashSet<>(broker.leaderReplicas())) {
```

Comment:
This goal does not keep the replicas to move out limited to immigrant replicas in case there are dead brokers -- i.e. replicas whose original broker is not this one. Wouldn't lack of this prioritization cause leadership/replica actions across alive brokers during fixing broker failures?

Don't we have a test case to ensure that the moves are limited to replicas that came from failed brokers -- if not can we add a test case?